### PR TITLE
Update checkoutRequiredCodeToBuild() to checkout the expected branch

### PR DESCRIPTION
**Pull request by pshipton**
Using the expected branch is required for
https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/42 to work and automatically checkout the same branch of OpenJ9 and OMR.

### CODE REVIEW

File: sbin/prepareWorkspace.sh [hunk 1]
```diff
@@ -190,11 +190,16 @@ checkoutRequiredCodeToBuild() {
...
-            echo "Checked out origin/${BUILD_CONFIG[BRANCH]}"
+            git reset --hard "origin/${BUILD_CONFIG[BRANCH]}" || rc=$?
+            if [ $rc -eq 0 ]; then
+              echo "Checked out origin/${BUILD_CONFIG[BRANCH]}"
+            else
+              echo "Failed cmd: git reset --hard \"origin/${BUILD_CONFIG[BRANCH]}\""
+            fi
           else
-            echo "Failed cmd: git reset --hard \"origin/${BUILD_CONFIG[BRANCH]}\""
+            echo "Failed cmd: git checkout -B ${BUILD_CONFIG[BRANCH]} origin/${BUILD_CONFIG[BRANCH]}"
```
Comment (keithc-ca) [inline sbin/prepareWorkspace.sh:202]: This should include the quotes for consistency with other failure messages (e.g. lines 199, 208).

File: sbin/prepareWorkspace.sh [hunk 2]
```diff
@@ -190,11 +190,16 @@ checkoutRequiredCodeToBuild() {
         # shellcheck disable=SC2086
         git fetch --all ${BUILD_CONFIG[SHALLOW_CLONE_OPTION]} || rc=$?
         if [ $rc -eq 0 ]; then
-          git reset --hard "origin/${BUILD_CONFIG[BRANCH]}" || rc=$?
+          git checkout -B "${BUILD_CONFIG[BRANCH]}" "origin/${BUILD_CONFIG[BRANCH]}" || rc=$?
```
Comment (keithc-ca) [inline sbin/prepareWorkspace.sh:193]: Suggest a comment here would be helpful:
```
          # It's not sufficient to checkout the commit at the head of the
          # requested branch, the name of that branch may also be significant,
          # so create one with the right name and check it out.
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by pshipton** [comment]
Updated.

**Review by keithc-ca — APPROVED** [review]

**Comment by pshipton** [comment]
@AdamBrousseau this can be merged any time. If it causes a problem for the milestone/RC/GA build, we can revert or fix it.

**Review by AdamBrousseau — COMMENTED** [review]
Can we make a comment why the `checkout` should to be before the `reset`? It's not entirely clear to me so having it in writing may help if we need to look at it in the future. Also was wondering if we plan to upstream this to Adopt?

**Comment by keithc-ca** [comment]
I think the only effect of `reset --hard` was to check out the desired commit. The problem was that it operated on the default branch. The proposed order avoids modifying any branches (even those local to the clone), other than to possibly create the requested (local) branch (tracking the remote branch of the same name).

After running `checkout -B`, `reset --hard` should have no effect. Done in the reverse order might leave the (local) branch (with the default name) at a commit other than where the corresponding, remote tracking branch points in the remote repository. Since we never push that (default) branch to the remote repository, it never mattered.

Because `reset --hard` should never do anything after `checkout -B`, we _could_ remove `reset --hard`, but I think we might have a harder time convincing Adoptium of that.

The bottom line is that we (OpenJ9) care what the branch name is as the build proceeds, so we need to use some command that arranges that.

**Comment by AdamBrousseau** [comment]
Talked to Pete.
When we do the reset we are on some branch (openj9 usually). Resetting to the right SHA will not checkout that branch. So when the ext code sniffs for the branch name it won't return what we want.
We do a fresh clone every time so the reset shouldn't be needed after the checkout but leaving it in doesn't hurt. I'm still not entirely sure how it would be different if we did the reset first then the checkout. Adopt might have a comment and also they don't clean their WS between builds so they may hit that code differently.
